src/lib/support/BitMask.h

```diff
@@ -0,0 +1,134 @@
+
+    template <typename... Args>
+    BitMask(FlagsEnum flag, Args &&... args) : BitFlags<FlagsEnum, StorageType>(flag, std::forward<Args>(args)...)
+    {}
+
+    template <typename... Args>
+    BitMask(IntegerType value, Args &&... args) : BitFlags<FlagsEnum, StorageType>(value, std::forward<Args>(args)...)
+    {}
+
+    BitMask & operator=(const BitMask &) = default;
+
+    BitMask & operator=(const BitFlags<FlagsEnum, StorageType> & other)
+    {
+        BitFlags<FlagsEnum, StorageType>::SetRaw(other.Raw());
+        return *this;
+    }
+
+    BitMask & operator=(FlagsEnum value)
+    {
+        BitFlags<FlagsEnum, StorageType>::SetRaw(static_cast<StorageType>(value));
+        return *this;
+    }
+
+    /**
+     * GetField to value via a mask shifting.
+     *
+     * @param mask      Typed flag(s) to used as mask, flag(s) must be contiguous. Any flags not in @a v are unaffected.
+     * @returns         Value from the underlying field/mask
+     */
+    constexpr BitMask & SetField(FlagsEnum mask, IntegerType value)
+    {
+        IntegerType bitMask = static_cast<IntegerType>(mask);
+        IntegerType shift   = GetShiftToFirstSetBit(bitMask);
+
+        // Value should be fully contained within the shift mask.
+        assert((value & (mask >> shift)) == value);
+
+        // Clear bits overlayed by the mask
+        IntegerType updated = static_cast<IntegerType>(BitFlags<FlagsEnum, StorageType>::Raw() & ~bitMask);
+        // Set the right bits
+        updated |= static_cast<IntegerType>(bitMask & (value << shift));
+
+        BitFlags<FlagsEnum, StorageType>::SetRaw(updated);
+
+        return *this;
+    }
+
+    /**
+     * GetField to value via a mask shifting.
+     *
+     * @param mask      Typed flag(s) to used as mask, flag(s) must be contiguous. Any flags not in @a v are unaffected.
+     * @returns         Value from the underlying field/mask
+     */
+    IntegerType GetField(FlagsEnum mask) const
+    {
+        {
+            IntegerType bitMask = static_cast<IntegerType>(mask);
+            IntegerType shift   = GetShiftToFirstSetBit(bitMask);
+
+            // Forward the right bits
+            return static_cast<IntegerType>(((BitFlags<FlagsEnum, StorageType>::Raw() & bitMask) >> shift));
+        }
+    }
+
+protected:
+    /// Get the shift amount to reach the first non-zero bit in a bitmask
+    ///
+    /// Examples:
+    ///    GetShiftToFirstSetBit(0b0000001) == 0
+    ///    GetShiftToFirstSetBit(0b0011100) == 2
+    ///    GetShiftToFirstSetBit(0b0110000) == 4
+    ///
+    /// Note: This does NOT validate if the given mask is a valid mask. So:
+    ///    GetShiftToFirstSetBit(0b0100010) == 1
+    ///    GetShiftToFirstSetBit(0b010110100) == 2
+    static constexpr IntegerType GetShiftToFirstSetBit(IntegerType bitMask)
+    {
+        IntegerType count = 0;
+        while (!(bitMask & 0x1) && (count < std::numeric_limits<IntegerType>::max()))
```

Comment:
This does not seem like the right condition.  IntegerType is the type of the integer underlyin the bitmap, right?  So if that's uint32_t, say, this loop will keep going until `count` is 2^32 if the mask happens to be 0.  Then it will do pretty odd things when shifting by 2^32....

The limit here should likely be `std::numeric_limits<IntegerType>::digits()`.  And the return type of this function should probably not be IntegerType....

Reply:
this is a good catch I originally did this with the max() , indeed this should 32 iteration not 2^32.

Reply:
Updated logic to not using count as a limiting factor but rather shifting until reaching the right location.